components/zcash_address/src/lib.rs

```diff
@@ -1,3 +1,116 @@
+//! *Parser for all defined Zcash address types.*
+//!
+//! This crate implements address parsing as a two-phase process, built around the opaque
+//! [`ZcashAddress`] type.
+//!
+//! - [`ZcashAddress`] can be parsed from, and encoded to, strings.
+//! - [`ZcashAddress::convert`] or [`ZcashAddress::convert_if_network`] can be used to
+//!   convert a parsed address into custom types that implement the [`TryFromAddress`] or
+//!   [`TryFromRawAddress`] traits.
+//! - Custom types can be converted into a [`ZcashAddress`] via its implementation of the
+//!   [`ToAddress`] trait.
+//!
+//! ```text
+//!         s.parse()              .convert()
+//!         -------->              --------->
+//! Strings           ZcashAddress            Custom types
+//!         <--------              <---------
+//!         .encode()              ToAddress
+//! ```
+//!
+//! # Using this crate
+//!
+//! ## I just need to validate Zcash addresses
+//!
+//! ```
+//! # use zcash_address::ZcashAddress;
+//! fn is_valid_zcash_address(addr_string: &str) -> bool {
+//!     addr_string.parse::<ZcashAddress>().is_ok()
+//! }
+//! ```
```

Comment:
The critical thing that I think this documentation is lacking is clarity as to how `ZcashAddress` relates to the protocol types, specifically that if you want to actually _use_ the addresses and/or keys parsed using this crate, you need to use something like the types in `zcash_client_backend`. To that point, I think that we ought to factor those out into their own crate at this point; it shouldn't be necessary to pull in all of `zcash_client_backend` just to get access to the address types that you end up wanting to use with `zcash_primitives`. 

Alternately, we could promote the unified address types up into `zcash_primitives`.

Reply:
Ideally I'd want examples that use the `sapling-crypto`, `orchard` and/or `zcash_client_backend` types here, because that's what users will actually need.

Reply:
Indeed, the thing this crate is missing is another crate we can point people to when they are writing a Rust app and _don't_ need their own types, and are happy to use ours. This crate isn't intended to be that crate.

Reply:
Added a few more notes about this.